src/algorithms/synthesis/encoding.cpp

```diff
@@ -150,8 +161,46 @@ namespace syrec {
             const auto requiredConstants = outputSize - inputSize;
             auto       newCube           = input;
             newCube.reserve(outputSize);
-            for (std::size_t i = 0; i < requiredConstants; i++) {
-                newCube.insertZero();
+
+            const auto oldConstantVec         = tt.getConstants();
+            const auto currentConstantVecSize = oldConstantVec.size();
+
+            if (appendZero) {
+                for (std::size_t i = 0; i < requiredInConstants; i++) {
+                    // based on the requiredInConstants, zeros are appended to the inputs.
+                    newCube.emplace_back(false);
+                    if (currentConstantVecSize != nBits) {
+                        // reset and resize the constants.
+                        tt.resetConstants();
+                        tt.resizeConstants(currentConstantVecSize + 1);
+                        for (auto cg = 0U; cg < currentConstantVecSize; cg++) {
+                            // storing the old constant bits in proper order.
+                            if (oldConstantVec[cg]) {
+                                tt.setConstant(cg + 1);
+                            }
+                        }
+                        // setting the newly added constant bit.
+                        tt.setConstant(0);
+                    }
+                }
+            } else {
+                // based on the requiredConstants, zeros are inserted to the inputs.
+                for (std::size_t i = 0; i < requiredConstants; i++) {
+                    newCube.insertZero();
+                    if (currentConstantVecSize != nBits) {
+                        // reset and resize the constants.
+                        tt.resetConstants();
+                        tt.resizeConstants(currentConstantVecSize + 1);
+                        for (auto cg = 0U; cg < currentConstantVecSize; cg++) {
+                            if (oldConstantVec[cg]) {
+                                // storing the old constant bits in proper order.
+                                tt.setConstant(cg);
+                            }
+                        }
+                        // setting the newly added constant bit.
+                        tt.setConstant(currentConstantVecSize);
+                    }
```

Comment:
Similar to above. This can be reduced to a resize and setting of the constant (without the if)